### ISSUE TITLE
Can't enforce 1 task at a time.

### ISSUE THREAD

**Issue by rs-garrick** [aws/containers-roadmap#959]
### Summary
I'm trying to enforce a strict "one task at a time" for a particular ECS service. The task requires a lengthy shutdown process (1 minute to an hour) before it can safely exit. Having 0 tasks for a short time is perfectly fine. It is vital that, for this particular service, no more than 1 task run at a time, and that it follows the potentially lengthy shutdown procedure. No matter what I do, ECS is starting new tasks too early.

### Description

On the service, I have "minimum healthy percent" set to 0, and "maximum percent" set to 100. This should ensure 1 task at a time, right?

The entrypoint script traps the SIGTERM to start the shutdown process and wait for it to complete. As such, I have a very long "stop timeout" parameter in my container definition. This method is working correctly.

Unfortunately, ECS is deregistering the task as soon as the task-stop is sent, and then starting a new task. I can easily observe the old container still running after the ECS task is deregistered. While the container is still running, the task is not visible in the ECS console. This results in overlapping tasks.

### Expected Behavior
With "maximum percent" set to 100, I expect the task to be fully and completely gone before the next task is started. I expect all running containers to be visible in the ECS console. I expect the task to be deregistered after the container has actually exited.

### Observed Behavior
Tasks are deregistered when the `docker stop` command is sent, but before the container actually exits. In my case, this could be an hour later.

In addition, the ECS host with the unregistered running container is "stuck". It can't start new tasks until it exits. This is OK when the container is expected to be exiting shortly, but not OK if it's going to be an hour wait.

### Environment Details
Amazon ECS Agent - v1.40.0 (17e8d834)
Amazon Linux release 2 (Karoo)
ECS_ENABLE_CONTAINER_METADATA=true
ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION=1h
ECS_CONTAINER_STOP_TIMEOUT=100m
ECS_ENABLE_UNTRACKED_IMAGE_CLEANUP=true

### Supporting Log Snippets
ECS-agent logs are perfectly normal. Tasks are transitioning their states appropriately. The problem is at the ECS service layer.

**Comment by sharanyad** [comment]
related: https://github.com/aws/containers-roadmap/issues/325

**Comment by marsavela** [comment]
I just tested this and it seems it works now. Setting `minimum_healthy_percent = 0` and `maximum_percent = 100`, ECS firstly kills the running task and then starts the new one. 👍  Maybe someone else can test it too.

**Comment by qs-synth** [comment]
@marsavela, I've tested this with my service and can confirm the `minimum_healthy_percent = 0` and `maximum_percent = 100` settings are working now. There is a slight gap between the old service stopping and new service starting, but that is expected.

**Comment by RaJiska** [comment]
Hi @sharanyad , @marsavela m @qs-synth ,

Having the same issue with a replica of 3 tasks. The service is configured with the constraint `distinctInstance` and is running on a cluster of three VMs. With `minimum_healthy_percent = 50`, I'd expect tasks to be replaced one by one, however, as soon as one of the task is set to `DEREGISTERING`, a new one gets scheduled instead of staying in a `PROVISIONING` state pending the task to be fully killed.

**Comment by drspacemanphd** [comment]
> Hi @sharanyad , @marsavela m @qs-synth ,
> 
> Having the same issue with a replica of 3 tasks. The service is configured with the constraint `distinctInstance` and is running on a cluster of three VMs. With `minimum_healthy_percent = 50`, I'd expect tasks to be replaced one by one, however, as soon as one of the task is set to `DEREGISTERING`, a new one gets scheduled instead of staying in a `PROVISIONING` state pending the task to be fully killed.

Hi @sharanyad, I am also seeing similar behavior for a service whether its deployed as a `REPLICA` or `DAEMON` service. As an example, for the `DAEMON` service configuration, I have 2 container instances as part of an ASG, with a `minimum_healthy_percent` of 50. As such, I would expect one task to be fully stopped before the new task is started on each instance. However, I can clearly see that the updated task is started while the old task has a lastStatus of `DEPROVISIONING`, and the underlying container still running. This is causing errors because tasks are configured to bind to host ports, and port binding conflicts arise if a new task is started while the container of the old task is still running.

I guess part of my confusion is that, given the [official docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/update-service.html), I can actually see this being expected behavior. When I configure my service to be of `REPLICA` type, I set the `minimum_healthy_percent` to be 50 and the `maximum_healthy_percent` of 100, I see the exact same behavior and errors as above. According to the docs,

>If a service is using the rolling update (ECS) deployment type, the maximum percent parameter represents an upper limit on the number of tasks in a service that are allowed in the RUNNING or PENDING state during a deployment, as a percentage of the desired number of tasks (rounded down to the nearest integer).

Given the above, I can see how this could be expected in the following way:
1. With a minimum of 50 and a maximum of 100, the service must first signal to de-provision the task.
2. With a state of `DEPROVISIONING`, the percentage of healthy tasks is now 50%. Given that the deployment maximum is 100, and `DEPROVISIONING` tasks do not count towards the maximum healthy percent, a new task is started.
3. The underlying container of the `DEPROVISIONING` task is still running as the new task is started. The port bindings clash and an error is issued.

I'm happy to try reproducing with a very basic express API image and provide more details/ecs agent logs.
